### ISSUE TITLE
API is confusing

### ISSUE THREAD

**Issue by AnonN10** [KhronosGroup/KTX-Software#603]
In the documentation I found that there are multiple structures for KTX textures: `ktxTexture` and `ktxTexture2`, I presume they represent KTX and KTX2? Then I suppose I should be using `ktxTexture2` since it's the newest one, that much seems logical. Now I need to serialize the texture, for that the documentation states there is a method `ktxTexture2_WriteToMemory`, I read how it works and try to write the code that uses it, but then there is no such function in my copy of `ktx.h` for some reason. I begin looking at the documentation again and find that there's also `ktxTexture1_WriteKTX2ToMemory`, what does it do exactly? Okay, being very unsure what to do I open up the header and look for `ktxTexture2_WriteToMemory` and this is what I find:
```
/**
 * @~English
 * @brief Helper for calling the WriteToMemory virtual method of a ktxTexture.
 * @copydoc ktxTexture2_WriteToMemory
 */
#define ktxTexture_WriteToMemory(This, ppDstBytes, pSize) \
                  (This)->vtbl->WriteToMemory(This, ppDstBytes, pSize)
```
Turns out I should be using this then, right? Why is the documentation so bad and out of sync with the code? In one of the releases.
You can simply close this issue if my guess that I should use `ktxTexture_WriteToMemory` is correct, otherwise I want to know how to properly use this library (and its docs).

**Comment by MarkCallow** [comment]
There is a class hierarchy. `ktxTexture` is the base class with stuff common to both KTX v1 and v2. `ktxTexture1` is a derived class for v1 textures and `ktxTexture2` is for  v2 textures. Because it is a C API implementing a class hierarchy has a few rough edges. As the comment for the macro says, it is a helper for calling a virtual method. To write your texture to memory just do

```
ktxTexture_WriteToMemory((ktxTexture*)yourTexture, &dstBytes, &size);
```

This will call the underlying, but not exposed in `ktx.h`, `ktxTexture1_WriteToMemory` or `ktxTexture2_WriteToMemory` as appropriate for `yourTexture`. Your guess is correct. Which documentation says to use `ktxTexture2_WriteToMemory`?

Did you look at the [complete documentation](https://github.khronos.org/KTX-Software/libktx/index.html) or just what is in `ktx.h`?

The documentation is not out of sync with the code (error are always possible of course). It is generated by Doxygen from comments in the source code along side the code being documented and it is updated with the code. Unfortunately Doxygen is designed for documenting source code not APIs so the overall structure of the doc is less than ideal.

**Comment by AnonN10** [comment]
> Did you look at the [complete documentation](https://github.khronos.org/KTX-Software/libktx/index.html) or just what is in `ktx.h`?

Indeed, I mean this documentation whenever I mention it in the first post. Seems like you have an error, you probably didn't pay too much attention to what I wrote, there is a mismatch:
` * @copydoc ktxTexture2_WriteToMemory`
for the function:
`#define ktxTexture_WriteToMemory(This, ppDstBytes, pSize) \`
this needs to be corrected, as otherwise currently the docs say this method exists: https://github.khronos.org/KTX-Software/libktx/group__writer.html#ga0113194b669b789a22d85b7ad992f117 but it really doesn't. I think there were other cases like this where functions mentioned by the docs weren't present in `ktx.h`, but I forgot which ones, sorry. And thanks for answering.

**Comment by MarkCallow** [comment]
I've removed the use of `ktxTexture2_WriteTo...` in the "Usage Overview". Unfortunately if I make the documentation for `ktxTexture2_WriteToMemory` etc. `@internal` they are no longer found by `@copydoc` so I cannot hide them from the generated docs. I do not want to have to duplicate this documentation in `ktx.h` nor have it only in `ktx.h` where it will be far from the implementing code.

**Comment by AnonN10** [comment]
@MarkCallow could you also change another part of "Usage Overview" to use newer API as well?
Update this part
```
int quality = 100;
result = ktxTexture2_CompressBasis(texture, quality);
```
to use function `ktxTexture2_CompressBasisEx` instead. Because `ktxTexture2_CompressBasis` always produces exaggerated blocky artifacts.

**Comment by MarkCallow** [comment]
Actually `ktxTexture2_CompressBasis` calls `ktxTexture2_CompressBasisEx`. The difference must be from the `quality` value you are supplying. Are you using 100? The default value, which you get with `ktxTexture2_CompressBasisEx` if you don't set the `quality` field in its `params` structure, is 128. Passing the value zero to `ktxTexture2_CompressBasis` is the will get the default value of 128 or you can explicitly specify 128. I will update the example but will continue to use `ktxTexture2_CompressBasis`.

Thanks for pointing out the problem.

**Comment by AnonN10** [comment]
I think the way to get rid of the artifacts is actually to set `ktxBasisParams::uastc` to `true`, from my understanding it's set to false by default, and UASTC is a superior encoding.

**Comment by MarkCallow** [comment]
> I think the way to get rid of the artifacts is actually to set ktxBasisParams::uastc to true

That will do it at the expense of a larger file. You need to use `ktxTexture2_DeflateZstd` on the result and possibly experiment RDO values to reduce the size down a bit.

When `ktxTexture2_CompressBasis` was added ETC1S/BasisLZ compression was the only scheme available so it had to continue to use that scheme. I'll make sure to talk about UASTC in the 'Usage Overview' update.

**Comment by AnonN10** [comment]
I just noticed that my main complaint is still relevant, the comment that doxygen picks up to make the documentation still refers to `ktxTexture_WriteToMemory` as `ktxTexture2_WriteToMemory` which doesn't exist in `ktx.h` file, yet makes it look like it does.

**Comment by MarkCallow** [comment]
Are you referring to the comment before the definition of `ktxTexture_WriteToMemory` in `ktx.h`? I looked into fixing this before. I did pay attention to what you wrote. Unfortunately the only way to fix it is to put the complete documentation before the macro definition in ktx.h. I prefer to keep the documentation with the implementation, which is one of `ktxTexture[12]_WriteToMemory`, to make it more certain it will be updated when the implementation is and to not bloat the public header file. I tried to mark those functions as `@internal` so they wouldn't appear in the docs but then the `@copydoc` fails.
